cc/position.cc

```diff
@@ -424,6 +429,264 @@ float Position::CalculateScore(float komi) {
   return static_cast<float>(score) - komi;
 }
 
+// A _region_ is a connected set of intersections regardless of color.
+// A _black-enclosed region_ is a maximum region containig no black stones.
+// A black-enclosed region is _small_ if all of its empty intersections are
+// liberties of the enclosing black stones.
+// A small black-enclosed region is _vital_ to an enclosing chain if all of its
+// empty intersections are liberties of that chain. Note that a small
+// black-enclosed region may not be vital to any of the enclosing chains. For
+// example:
+//   . . . . . .
+//   . . X X . .
+//   . X . . X .
+//   . X . . X .
+//   . . X X . .
+//   . . . . . .
+//
+// A set of black chains X is _unconditionally alive_ if each chain in X has at
+// least two distinct small black-enclosed regions that are vital to it.
+// A region enclosed by set of unconitionally alive black chains is an
+// unconditionally alive black region.
+//
+// Given these definitions, Benson's Algorithm finds the set of unconditionally
+// alive black regions as follows:
+//  - Let X be the set of all black chains.
+//  - Let R be the set of small black-enclosed regions of X.
+//  - Iterate the following two steps until neither one removes an item:
+//    - Remove from X all black chains with fewer than two vital
+//      black-enclosed regions in R.
+//    - Remove from R all black-enclosed regions with a surrounding stone in a
+//      chain not in X.
+//
+// Unconditionally alive chains are also called pass-alive because they cannot
+// be captured by the opponent even if that player always passes on their turn.
+// More details:
+//   https://senseis.xmp.net/?BensonsDefinitionOfUnconditionalLife
+std::array<Color, kN * kN> Position::CalculatePassAliveRegions(
+    Color color) const {
+  struct BensonGroup {
+    Coord first_stone = Coord::kInvalid;
+    std::set<Coord> liberties;
+    int num_vital_regions = 0;
+    bool is_pass_alive = false;
+  };
+
+  struct BensonRegion {
+    Coord first_stone = Coord::kInvalid;
+    std::set<Coord> empty_points;
+    std::set<Coord> other_color_points;
+    std::set<BensonGroup*> vital_for;
+    std::set<BensonGroup*> enclosing_groups;
+    bool is_pass_alive = false;
+  };
+
+  std::vector<std::unique_ptr<BensonRegion>> region_pool;
+  std::vector<std::unique_ptr<BensonGroup>> group_pool;
+  std::array<BensonRegion*, kN * kN> regions;
+  std::array<BensonGroup*, kN * kN> groups;
+
+  // Build the set of all groups.
+  board_visitor_->Begin();
+  for (int row = 0; row < kN; ++row) {
+    for (int col = 0; col < kN; ++col) {
+      Coord c(row, col);
+      if (stones_[c].color() != color || !board_visitor_->Visit(c)) {
```

Comment:
trying to understand this visitor pattern. The outer double-for loop guarantees that every square is visited at least once. But is every square visited exactly once? Otherwise I feel like double-counting would occur when compiling regions/neighboring groups.

Reply:
The `BoardVisitor` guarantees that each stone is only visited once. The `Visit` method returns `true` the first time a point is visited and `false` otherwise. `Visit` pushes the `Coord` of the point onto an internal queue, which `Next` pops from.

Here's a simple example of using just the `BoardVisitor` to visit every point on the board:

```C++
   BoardVisitor bv;
   bv.Begin()
   bv.Visit(Coord::FromGtp("A1"));
   while (!bv.Done()) {
     Coord coord = bv.Next();
     std::cout << "Visited " << coord << "\n";
     for (auto neighbor_coord : GetNeighbors(coord)) {
       bv.Visit(neighbor_coord);
     }
   }
```